### PR TITLE
Rtl labels support

### PR DESCRIPTION
**Pull request by hodbauer**
reference: #2543
For usage see Label documentation and Label sandcastle example.
Supports new lines
Currently supports Hebrew (should easily support other RTL languages)
Tasks:
- [ ] Add support for more languages
- [ ] Fix case when there is RTL characters, numbers and space (ex: `משתמש10 `)

### CODE REVIEW

**Comment by pjcozzi** [comment]
Wow, very cool thanks for the nice contribution @hodbauer and @YonatanKra!

@ggetz could you take a first look at this then pass it on to @bagnell for a final review and merge?

File: Source/Scene/Label.js [hunk 1]
```diff
@@ -1030,6 +1044,42 @@ define([
...
+         * myLabelEntity.rtl = true;
+         * myLabelEntity.text = 'טקסט חדש'
+         */
+        rtl : {
+            get : function() {
+                return this._rtl;
+            },
+            set : function(value) {
+                if (this._rtl !== value) {
+                    this._rtl = value;
```
Comment (ggetz) [inline Source/Scene/Label.js]: I think you should call `reverseRtl` here in addition to when setting the text.

File: Source/Scene/Label.js [hunk 2]
```diff
@@ -1196,5 +1246,171 @@ define([
...
+                return '<';
+        }
+    }
+
+    /**
+     *
+     * @param {String} text the text to parse and reorder
+     * @returns {String} the text as rtl direction
+     */
+    Label.prototype.reverseRtl = function(value) {
```
Comment (ggetz) [inline Source/Scene/Label.js]: This function should be private, I don't think anyone should be calling this directly.

File: Specs/Scene/LabelCollectionSpec.js [hunk 3]
```diff
@@ -1840,6 +1840,43 @@ defineSuite([
...
+               rtl : true
+           });
+           scene.renderForSpecs();
+
+           expect(label.rtl).toEqual(true);
+           expect(label.text).toEqual(text);
+        });
+
+        it('should reverse text when there is only hebrew characters and rtl is true', function() {
+           var text = 'שלום';
```
Comment (ggetz) [inline Specs/Scene/LabelCollectionSpec.js]: Can you add an additional tests that cover different scenarios like spaces and brackets in words?

File: Source/Scene/Label.js [hunk 4]
```diff
@@ -1196,5 +1246,171 @@ define([
...
+                    }
+                }
+                else if (subText.Type === types.RTL) {
+                        result = spliceWord(result, splicePointer, reverseWord(subText.Word));
+                    }
+                    else if (subText.Type === types.LTR) {
+                        result += subText.Word;
+                        splicePointer = result.length;
+                    }
+                    else if (subText.Type === types.WEAK || subText.Type ===types.BRACKETS) {
```
Comment (ggetz) [inline Source/Scene/Label.js]: spacing after `===`

Comment (ggetz) [inline Source/Scene/Label.js]: and on line 1354

File: Source/Scene/Label.js [hunk 5]
```diff
@@ -1196,5 +1246,171 @@ define([
...
+                    if (subText.Type === types.RTL) {
+                        result = reverseWord(subText.Word) + result;
+                        splicePointer = 0;
+                    }
+                    else if (subText.Type === types.LTR) {
+                        result = spliceWord(result,splicePointer, subText.Word);
+                        splicePointer += subText.Word.length;
+                    }
+                    else if (subText.Type === types.WEAK || subText.Type ===types.BRACKETS) {
+                        if (parsedText[wordIndex -1].Type === types.RTL) {
```
Comment (ggetz) [inline Source/Scene/Label.js]: Add a space after the operator here and throughout.

File: Source/Scene/Label.js (hunk 1 above)
Comment (hodbauer) [inline Source/Scene/Label.js]: Thank you @ggetz for your comments. I will fix them after the weekend.
I have a specific question about this comment. Why to call `reverseRtl` here? If the text is written in rtl language, I don't see a reason to change the rtl flag while using it, the algorithm will work good even for ltr (left to right) languages.

Comment (ggetz) [reply Source/Scene/Label.js]: Thanks @hodbauer!

If I understand correctly, if `rtl` is set to false, and you update `text` to an rtl language, then set `rtl` to true, it won't display correctly:

```
label.text = 'שלום';
label.rtl = true;
```

That may be confusing to people using this feature.

**Comment by ggetz** [comment]
Hi @hodbauer, I see you made updates. Are you ready for another look?

**Comment by hodbauer** [comment]
Hi @ggetz 
I'll be glad if you go over it again.
I want to clarify why I didn't change the logic at the `rtl` setter. From a check that I've done it seems that the label ran the algorithm even if I didn't call the `reverseRtl` from the setter of `rtl`.

**Comment by ggetz** [comment]
@hodbauer, everything looks good except I still think you should reconsider the `rtl` setter.

In the sandcastle example, if I run this:
```
    var label = viewer.entities.add({
        position : Cesium.Cartesian3.fromDegrees(-75.1641667, 39.9522222),
        label : {
            text : 'Master (אדון): Hello\nתלמיד (student): שלום'
        }
    });
    
    label.rtl = true;
```
The rtl portion of the text will be reversed, which is not the same results as this:
```
viewer.entities.add({
        position : Cesium.Cartesian3.fromDegrees(-75.1641667, 39.9522222),
        label : {
            text : 'Master (אדון): Hello\nתלמיד (student): שלום',
            rtl : true
        }
    });
```

**Comment by hodbauer** [comment]
Ok @ggetz I'll check this case again and update the code according to the results

**Comment by hodbauer** [comment]
I have a question @ggetz.
At the sandcastle I ran a simple case like that:
```javascript
var label = viewer.entities.add({
        position : Cesium.Cartesian3.fromDegrees(-75.1641667, 39.9522222),
        label : {
            text : 'Hi'
        }
    });
    label.text = 'Bye';
```
I was expect that the label at Cesium will be `Bye`, but it was `Hi`.
If this case not work, how the case that you mention before will work for me? What I was missing?
(It also was checked at the official Sandcastle)

**Comment by hodbauer** [comment]
After another check at Sandcastle I've got the next solution.
```
    var entity = viewer.entities.add({
        position : Cesium.Cartesian3.fromDegrees(-75.1641667, 39.9522222),
        label : {
            text : 'Master (אדון): Hello\nתלמיד (student): שלום'
        }
    });
    entity.label.rtl = true;
```
now it just work fine.
Waiting for your response @ggetz.

**Comment by ggetz** [comment]
@hodbauer My mistake, you're correct!

@bagnell You're up for final review and merge.

**Comment by mramato** [comment]
Are we okay with using `rtl` rather than `rightToLeft`, we usually don't use abbreviations in our API, but rtl is a fairly common one.

**Comment by pjcozzi** [comment]
> Are we okay with using rtl rather than rightToLeft, we usually don't use abbreviations in our API, but rtl is a fairly common one.

Let's go with `rightToLeft`.  I do not think `rtl` is common enough for our users.

**Comment by hodbauer** [comment]
@pjcozzi,
I renamed it from `rtl` to `rightToLeft`.

**Comment by hpinkos** [comment]
@hodbauer please resolve the merge conflict with master when you have a chance.  Thanks!

**Comment by hodbauer** [comment]
@pjcozzi , @ggetz , @hpinkos any news about this PR?

**Comment by mramato** [comment]
@hodbauer it looks like there are some issues with the commits in this PR.  Can you merge in master and reset it properly so just your commits show up.  If so I'll try and get this into master for the next release.

**Comment by hodbauer** [comment]
@mramato, I've updated the branch so it will be only my changes, thanks for the attention.

**Comment by hodbauer** [comment]
@mramato any news?
what is the next version date to update the `CHANGES.md` file as needed?

**Comment by ggetz** [comment]
@hodbauer November 1 is the next release, version 1.39.

@pjcozzi Did we need this reviewed by @bagnell or is this ready to go?

**Comment by mramato** [comment]
Sorry for the delay @hodbauer 

> Add support for more languages

What does this mean exactly?  Does this code not work on all RTL languages?

Have we tried this on all browsers?

File: CONTRIBUTORS.md [hunk 6]
```diff
@@ -160,3 +163,5 @@ See [CONTRIBUTING.md](CONTRIBUTING.md) for details on how to contribute to Cesiu
 * [Cody Guldner](https://github.com/burn123)
 * [Florent Cayré](https://github.com/fcayre)
 * [Nacho Carnicero](https://github.com/nacho-carnicero)
+* [Yonatan Kra](https://github.com/yonatankra)
```
Comment (pjcozzi) [inline CONTRIBUTORS.md]: Did you mean to add yourself to this section as well as the **webiks** section above?

**Comment by pjcozzi** [comment]
> @pjcozzi Did we need this reviewed by @bagnell or is this ready to go?

@bagnell, @mramato, or @tfili could review this.

Please make sure that test coverage is high enough as there are a lot of branches.

**Comment by hodbauer** [comment]
> Add support for more languages
> What does this mean exactly? Does this code not work on all RTL languages?

In the algorithm that I've wrote there is a regex that detects if the character is of RTL language.
But to actually detect different languages the letters of the language should be added to the regex.
I know how to write in Hebrew so it support it right now, but if we want to support more languages like Arabic it should be added to the regex and actually to check if it's looks good after rendering.

> Have we tried this on all browsers?

No, checked only on chrome for windows.

> Please make sure that test coverage is high enough as there are a lot of branches.

Is there a way to see what exactly covered?

I'll update the .md files in the coming days.

**Comment by pjcozzi** [comment]
For test coverage, see https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/TestingGuide#run-all-tests-against-combined-file-run-all-tests-against-combined-file-with-debug-code-removed

**Comment by hodbauer** [comment]
**summary after another iteration:**
> Have we tried this on all browsers?

It tested on Chrome, firefox and edge

> Please make sure that test coverage is high enough as there are a lot of branches.

hopefully all the branches covered by adding more tests. When Iv'e added the tests I fix some edge cases.
Another interesting thing that I found during this tests is that JScoverage lib not support hebrew characters, so I changed my regex that detect Hebrew from `/[א-ת]/` to the unicode represent `/[\u05D0-\u05EA]/`.
Also I was tried to understand if I can support Arabic or other rtl languages and understand that if I cannot read it I shouldn't promise that it work fine...

**Comment by pjcozzi** [comment]
@hodbauer thanks for the diligent updates.

@mramato did you want to look at this again and merge when ready?

**Comment by mramato** [comment]
Sure, I'll try and get to this before the end of the week.

File: Source/Scene/Label.js [hunk 7]
```diff
@@ -1196,5 +1249,184 @@ define([
...
+    }
+
+    /**
+     *
+     * @param {String} value the text to parse and reorder
+     * @returns {String} the text as rightToLeft direction
+     * @private
+     */
+    function reverseRtl(value) {
+        var rtlChars = /[\u05D0-\u05EA]/;
```
Comment (mramato) [inline Source/Scene/Label.js]: So if I'm understanding this code correctly, this regex detects whether the letter is RTL (in this case specifically Hebrew) but otherwise the rest of the code is the same for other languages?  So if we update this regex for other known RTL character sets (perhaps thise discussed [here](https://stackoverflow.com/questions/12006095/javascript-how-to-check-if-character-is-rtl) and [here](https://stackoverflow.com/questions/4330951/how-to-detect-whether-a-character-belongs-to-a-right-to-left-language)  Is that correct?  If so that's awesome and we can easily update this to handle the other languages.

File: Source/DataSources/LabelGraphics.js [hunk 8]
```diff
@@ -318,7 +321,15 @@ define([
...
-        disableDepthTestDistance : createPropertyDescriptor('disableDepthTestDistance')
+        disableDepthTestDistance : createPropertyDescriptor('disableDepthTestDistance'),
+
+        /**
+         * Gets or sets the ability to modify text characters direction.
+         * @memberof LabelGraphics.prototype
+         * @type {Property}
+         * @default false
+         */
+        rightToLeft: createPropertyDescriptor('rightToLeft')
```
Comment (mramato) [inline Source/DataSources/LabelGraphics.js]: This is basically an optimization, right?  If the string doesn't contain RTL characters it is a no-op?

File: Source/Scene/Label.js [hunk 9]
```diff
@@ -1196,5 +1249,184 @@ define([
         return false;
     };
 
+    function declareTypes() {
+        var TextTypes = {
```
Comment (mramato) [inline Source/Scene/Label.js]: I would just declare this at the top of the file instead of creating it every time.  Then just use `TextTypes` everywhere

**Comment by mramato** [comment]
I merged in master for you and just have those few suggestions/questions.

Thanks again for the awesome work here, I'm sure a huge section of our community will be excited to have RTL support at last.

File: Source/Scene/Label.js (hunk 7 above)
Comment (hodbauer) [reply Source/Scene/Label.js]: > but otherwise the rest of the code is the same for other languages? 

yes. it only necessary when writing a RTL language in Cesium. 

> we can easily update this to handle the other languages.

basically yes, I do not do that, because i cannot read other RTL languages.

File: Source/DataSources/LabelGraphics.js (hunk 8 above)
Comment (hodbauer) [reply Source/DataSources/LabelGraphics.js]: yes

File: Source/Scene/Label.js (hunk 9 above)
Comment (hodbauer) [reply Source/Scene/Label.js]: @mramato If I understand you, you want me to change the code to something like:
```js
var TextTypes = freezeObject({
    LTR : 0,
    RTL : 1,
    WEAK : 2,
    BRACKETS : 3
});
return TextTypes;
```
and put it inside the define near to the glyph functions as is (without a function scope)?

**Comment by hodbauer** [comment]
I don't know why but I've got this error while running travis ci:
```
Core/WebGLConstants
 ✗ freezes render states
	Expected function to throw an exception.
```
Any help about this issue, it not part of my changes

**Comment by ggetz** [comment]
@hodbauer Seems to be failing across multiple branches, I opened #5924

**Comment by hpinkos** [comment]
@hodbauer I fixed that test in master, so if you merge that into your branch CI should pass. Thanks!

**Comment by hodbauer** [comment]
@ggetz @hpinkos @mramato thanks for solving the problem!
@mramato we could continue with the PR

File: Source/DataSources/LabelGraphics.js (hunk 8 above)
Comment (mramato) [reply Source/DataSources/LabelGraphics.js]: Since this is just an optimization, how do you feel about having a single static `Label.enableRightToLeftDetection` variable that an application sets once globally rather than having to set it on every label?  It will save memory when there are many labels and I doubt it will have a performance impact in RTL apps since I would expect most of the labels to be RTL in that case.  Thoughts?

File: Source/Scene/Label.js (hunk 7 above)
Comment (mramato) [reply Source/Scene/Label.js]: Cool, we can merge this with just Hebrew and then I'll open a second PR with an updated regex to support additional languages.

File: Source/Scene/Label.js (hunk 9 above)
Comment (mramato) [reply Source/Scene/Label.js]: Yes, though you won't need the `return` statement.  You can just put the declaration at the top of the file (above `rebindAllGlyphs`), it will be scoped to Label.js.

File: Source/DataSources/LabelGraphics.js (hunk 8 above)
Comment (hodbauer) [reply Source/DataSources/LabelGraphics.js]: Sounds very interesting.
I'm not sure 100% how to implement it, but if it is the right way then I should do it as you offered

Comment (hodbauer) [reply Source/DataSources/LabelGraphics.js]: after a little bit of thinking I understand how to implement it, it will clean a lot of the work and it is nice.
my question is: how to declare the static variable so each application could manage it?

**Comment by hodbauer** [comment]
As @mramato offered, I changed the way to enable the right-to-left algorithm.
The new way is by setting a static parameter of `Label`, currently named `Label.enableRightToLeftDetection` to `true`, then every update to any label will run the algorithm.
The only issue about this as you can see is that the tests failed.
for example:
```js
            Label.enableRightToLeftDetection = true;
            var text = 'שלום';
            var label = labels.add({
                text : text
            });

            scene.renderForSpecs();

            expect(label.text).not.toEqual(text);
            expect(label.text).toEqual(text.split('').reverse().join(''));
```
This test failed although I enable the right-to-left algorithm. moreover, when I've checked the sandecastle it worked as expected.
So, how to use this parameter during tests?

File: Source/DataSources/LabelGraphics.js (hunk 8 above)
Comment (mramato) [reply Source/DataSources/LabelGraphics.js]: Just add a property directly on the `Label` object.  Something like this should work.

```javascript
   /**
     * Enabled detection and support of right to left text.
     *
     * @type {Boolean}
     * @default false
     *
     * @example
     * //Set once at the start of the application
     * Cesium.Label.enableRightToLeftDetection = true;
     * 
     * var myLabelEntity = viewer.entities.add({
     *   label: {
     *     id: 'my label',
     *     text: 'זה טקסט בעברית \n ועכשיו יורדים שורה'
     *   }
     * });
     */
    Label.enableRightToLeftDetection = false;
```

This will show up as static in the documentation and users just set `Cesium.Label.enableRightToLeftDetection = true` once in their app to use it. In your own code, you would just check `Label.enableRightToLeftDetection` directly as well.

Thanks.

**Comment by mramato** [comment]
> So, how to use this parameter during tests?

It looks like your code should be activating it, so I'm not sure what's going on.  I can dig in a little further and let you know what I find.

**Comment by hodbauer** [comment]
The code was OK, the problem was that during creation of a Label nothing calls to the text setter, so nothing run the algorithm.
The mandatory change that I've done at the last commit was to call the text setter in the constructor:
```js
this.text = defaultValue(options.text, '');
```
my friend @YonatanKra that help me with this PR done it before. and me, by mistake, removed it in the last changes.

File: Source/Scene/Label.js [hunk 10]
```diff
@@ -279,6 +290,15 @@ define([
                 }
                 //>>includeEnd('debug');
 
+                if (Label.enableRightToLeftDetection) {
+                    if (this._originalValue === value) {
+                        value = this._text;
+                        return;
+                    }
+                    this._originalValue = value;
+                    value = reverseRtl(value);
```
Comment (mramato) [inline Source/Scene/Label.js]: I just noticed that we are mutating the value supplied by the user so that the getter is actually the reversed text.  I think this is undesired behavior.  With few exceptions, a simple setter/getter should always return the value assigned to it.  I can submit a change to prevent this, but if you feel strongly about it, please let me know.

**Comment by mramato** [comment]
Awesome work here @hodbauer and @YonatanKra!  I submitted the tweak I mentioned, but otherwise I think this is ready so I'm going to merge.

If you find any issues or I accidentally broke something with my tweaks, please let me know.

Thanks again!
